calico/openstack/t_etcd.py

```diff
@@ -41,379 +48,99 @@
-                            # unnecessarily write out its (unchanged) data again below.
-                            LOG.debug("Existing etcd profile data is correct")
-                            correct_profiles.add(profile_id)
-                    else:
-                        # We don't want this profile any more, so delete the key.
-                        LOG.debug("Existing etcd profile key is now invalid")
-                        profile_key = key_for_profile(profile_id)
-                        self.client.delete(profile_key, recursive=True)
-                except etcd.EtcdKeyNotFound:
-                    LOG.info("Etcd data appears to have been reset")
-
-        # Now write etcd data for each profile that we need and that we don't
-        # already know to be correct.
-        for profile_id in self.needed_profiles.difference(correct_profiles):
-            self.write_profile_to_etcd(profile_id)
-
-    def write_profile_to_etcd(self, profile_id):
-        self.client.write(key_for_profile_rules(profile_id),
-                          json.dumps(self.profile_rules(profile_id)))
-        self.client.write(key_for_profile_tags(profile_id),
-                          json.dumps(self.profile_tags(profile_id)))
-
-    def profile_rules(self, profile_id):
-        inbound = []
-        outbound = []
-        for sgid in self.profile_tags(profile_id):
-            # Be tolerant of a security group not being here. Allow up to 20
-            # attempts to get it, waiting a few hundred ms in between: we might
-            # just be racing slightly ahead of a security group update.
-            rules = None
-            retries = 20
-            while rules is None:
-                try:
-                    with self._sgs_semaphore:
-                        rules = self.sgs[sgid]['security_group_rules']
-                except KeyError:
-                    LOG.warning("Missing info for SG %s: waiting.", sgid)
-                    retries -= 1
-
-                    if not retries:
-                        LOG.error("Gave up waiting for SG %s", sgid)
-                        raise
-
-                    # Wait for 200ms
-                    eventlet.sleep(0.2)
-
-
-            for rule in rules:
-                LOG.info("Neutron rule  %s : %s", profile_id, rule)
-                etcd_rule = _neutron_rule_to_etcd_rule(rule)
-                if rule['direction'] == 'ingress':
-                    inbound.append(etcd_rule)
-                else:
-                    outbound.append(etcd_rule)
-
-        return {'inbound_rules': inbound, 'outbound_rules': outbound}
-
-    def profile_tags(self, profile_id):
-        return profile_id.split('_')
+        # Elector, for performing leader election.
+        # TODO: This doesn't handle forking yet, we need to mix the PID in here
+        # somewhere.
+        self.elector = Elector(
+            client=self.client,
+            server_id=cfg.CONF.calico.elector_name,
```

Comment:
Surely this doesn't work if you have multiple API workers since they'll share config?  It needs to include the PID, at least (or just make it a UUID).